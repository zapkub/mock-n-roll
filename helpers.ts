import {
  SourceFile,
  TypeNode,
  SyntaxKind,
  MethodSignature,
  FunctionTypeNode,
  MethodDeclaration
} from 'ts-morph'
import {isEqual} from 'lodash';

export function getIdentifierListOfMethodArgs(
  method: MethodSignature | FunctionTypeNode | MethodDeclaration
) {
  return method
    .getParameters()
    .map(p => {
      if (p.getDotDotDotToken()) {
        return `...${p.getName()}`
      }
      return p.getName()
    })
    .join(',')
}

export function ImportTypeIfNeeded(
  src: SourceFile,
  mock: SourceFile,
  typeNode: TypeNode
) {
  typeNode.getChildrenOfKind(SyntaxKind.Identifier).forEach(t => {
    ImportByName(src, mock, t.getFullText().trim())
  })
  typeNode.getChildrenOfKind(SyntaxKind.ArrayType).forEach(t => {
    ImportTypeIfNeeded(src, mock, t)
  })
  typeNode.getChildrenOfKind(SyntaxKind.UnionType).forEach(t => {
    ImportTypeIfNeeded(src, mock, t)
  })
  typeNode.getChildrenOfKind(SyntaxKind.TypeReference).forEach(t => {
    ImportTypeIfNeeded(src, mock, t)
  })
  typeNode.getChildrenOfKind(SyntaxKind.SyntaxList).forEach(s => {
    s.getChildrenOfKind(SyntaxKind.PropertySignature).forEach(sig => {
      const ref = sig.getFirstChildByKind(SyntaxKind.TypeReference)
      if (ref) {
        ImportByName(src, mock, ref.getTypeName().getText())
      }
    })
  })
}

export function ImportByName(
  src: SourceFile,
  mock: SourceFile,
  name: string
): boolean {
  if (name.match(/^Pick$|^Promise$/)) {
    console.log(`skip to import ${name}`)
    return false
  }
  name = name.trim()

  const srcImportLine = src.getImportDeclarations().find(declar => {
    return !!declar.getNamedImports().find(named => named.getName() === name)
  })
  let moduleSpecifier: string | undefined
  if (srcImportLine) {
    if (srcImportLine.getModuleSpecifierSourceFile()) {
      moduleSpecifier = mock.getRelativePathAsModuleSpecifierTo(
        srcImportLine.getModuleSpecifierSourceFileOrThrow()
      )
    } else {
      moduleSpecifier = srcImportLine.getModuleSpecifierValue()
    }
  } else {
    const exp = src.getExportSymbols().find(ex => {
      return ex.getName() === name
    })
    if (exp) {
      moduleSpecifier = mock.getRelativePathAsModuleSpecifierTo(src)
    }
  }

  if (!moduleSpecifier) {
    console.log(`${name} not found in any reference`)
    return false
  }

  let importStatement = mock.getImportDeclarations().find(importDeclar => {
    const f = importDeclar.getModuleSpecifier().getLiteralText()
    return f === moduleSpecifier
  })
  if (importStatement) {
    if (
      !importStatement
        .getNamedImports()
        .find(named => named.getSymbolOrThrow().getName() === name)
    ) {
      importStatement.addNamedImport(name)
    }
    return true
  } else {
    mock.addImportDeclaration({
      namedImports: [name],
      moduleSpecifier
    })
    return true
  }
}

export const Any = () => {
  return '__MOCK__MATCH_ANY_THING'
}

var serialize = require('node-serialize')
export const pushCalled = (calls: any[], methodName: string, ...args: any[]) => {
  return (returnValue) => {
    calls.push([[methodName, serialize.serialize(args)], returnValue])
  }
}

export const on = (calls: any[], methodName: string, ...args: any[]) => {
  const result = calls.find(c => {
    const call = [methodName, serialize.serialize(args)]
    return isEqual(c[0], call)
  })
  if (!result) {
    throw new Error(`call ${methodName} with ${JSON.stringify(serialize.serialize(...args))} does not exists`)
  }
  return result[1]
}
