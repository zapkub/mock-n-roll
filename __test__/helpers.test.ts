import { pushCalled, on } from '../helpers'

test("PushCalled test", () => {


  const called = []


  pushCalled(called, "findSomething", function(a) {})("TEST")
  expect(called).toMatchSnapshot()
  expect(on(called, "findSomething", function(a) {})).toEqual("TEST")

  pushCalled(called, "findSomething", (a) => {}, () => "x")("TEST_2")
  expect(called).toMatchSnapshot()


})