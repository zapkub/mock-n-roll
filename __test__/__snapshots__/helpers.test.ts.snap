// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PushCalled test 1`] = `
Array [
  Array [
    Array [
      "findSomething",
      "{\\"0\\":\\"_$$ND_FUNC$$_function (a) { }\\"}",
    ],
    "TEST",
  ],
]
`;

exports[`PushCalled test 2`] = `
Array [
  Array [
    Array [
      "findSomething",
      "{\\"0\\":\\"_$$ND_FUNC$$_function (a) { }\\"}",
    ],
    "TEST",
  ],
  Array [
    Array [
      "findSomething",
      "{\\"0\\":\\"_$$ND_FUNC$$_(a) => { }\\",\\"1\\":\\"_$$ND_FUNC$$_() => \\\\\\"x\\\\\\"\\"}",
    ],
    "TEST_2",
  ],
]
`;
